### DISCUSSION TITLE
WebServer with AM2302 Sensor

### DISCUSSION THREAD

**Discussion by hasenradball** [earlephilhower/arduino-pico#2570] [General]
Dear @earlephilhower,

thank you for providing this library.

I checked this with the RaspberryPi PicoW in the following way, and observed a not expected behaviour.

- a) take as basis the example of "HelloServer"
- b) I added a single Sensor AM2302 (aka DHT22) by cpp library.
- c) changed the Website in this way that it shows the temperature and humidity values. 

When you use the sensor lib alone on the Pico/PicoW everything is fine:
sensor.read() ==> returns always Status == OK.

But when you use the Sensor lib in combination with the Webserver the 
sensor.read() ==> returns mostly not Status == OK.

What means Webserver + Sensor readings is too much for the PicoW, my assumption.
But it is difficult to believe for me.

Wanted to analyse more in detail.
Can you give some hints to this issue?

AM2302 Lib: [AM2302-Sensor](https://github.com/hasenradball/AM2302-Sensor)

**Comment by dquadros** [comment]
The response from DHT sensors is "time coded": the signal stays high for different times to indicate a 0 or a 1. The AM3202-Sensor library measure these times using a loop and delays, so it will be sensitive to interrupts. I observed similar problems with CircuitPython and a Adafruit library. The RP2040 has ways to do this more efficiently, by doing the reading in the other core or using the PIO.

**Comment by hasenradball** [comment]
See this issue on Arduino core and on pico SDK.

I will check this at ESP8266 but expectation is, this does not occur.

**Comment by hasenradball** [comment]
@dquadros have you examples shifting to other core?

**Comment by earlephilhower** [comment]
Run your sensor update in `setup1` and `loop1` like in [the example](https://github.com/earlephilhower/arduino-pico/blob/master/libraries/rp2040/examples/Multicore/Multicore.ino), updating a `volatile` global state variable, for example.

As @dquadros said, we may have more interrupts than the 8266 especially since the IP stack runs without needing `yield` to process and there's the USB stack with ~1000 IRQs/sec.  If the bitrate is high enough, you might be able to get away with disabling interrupts around the reading process.  But if it's a multi-millisecond procedure, I second his suggestion of using the PIO or another method to get the more precise timing.

**Comment by hasenradball** [comment]
Great Thanks for your feedback @earlephilhower + @dquadros .
Is there also an example of using the PIO?

**Comment by earlephilhower** [comment]
Moved to a discussion since it's not a core issue.  The PIO is a simple FSM that you can program with its own small assembly language.  See the [datasheet](https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf) where they show examples of things like Manchester decoding.

That said, it looks like you have a max of 80 bit times, each being ~100us on average.  So 8ms total during reception.  You might really want to try disabling interrupts around your readout process (no `delay` allowed without IRQs, though!) and see how that works.

**Comment by dquadros** [comment]
I have an example of PIO code, but for using with the SDK:

https://github.com/dquadros/PicoSensors/tree/main/SDK/Temperature/dht

**Comment by hasenradball** [comment]
@dquadros 
Thank you.
Did you wrote the *•pio code by yourself or was it generated?

**Comment by dquadros** [comment]
I wrote the pio code. To use it it has to be "assembled" by pioasm, this generates an include (.h) file, that is included in the cpp source.

Some adjustments need to be done to use this under Arduino.

**Comment by hasenradball** [comment]
Hi Guy,

checked wirh two cores, but issue exists. Not so much as with obly one core but is there.
Check with pio actually but first have to check the basics therefore.
